linked_list/linked_list.py

```diff
@@ -13,67 +13,126 @@ def __init__(self):
+                return True
+            cur = cur.next
+        return False
 
     # method that returns the length of the singly linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def length(self):
-        pass
+        cur = self.head
+        i = 0
+        while cur != None:
+            cur = cur.next
+            i += 1
+        return i
 
     # method that returns the value at a given index in the linked list
     # index count starts at 0
     # returns None if there are fewer nodes in the linked list than the index value
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def get_at_index(self, index):
-        pass
+        cur = self.head
+        for i in range(index):
+            if cur == None:
+                return None
+            cur = cur.next
+        return cur.value
+
 
     # method that returns the value of the last node in the linked list
     # returns None if the linked list is empty
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def get_last(self):
-        pass
+        if self.head == None:
+            return None
+        cur = self.head
+        prev = None
+        while cur != None:
+            prev = cur
+            cur = cur.next
+        return prev.value
 
     # method that inserts a given value as a new last node in the linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def add_last(self, value):
-        pass
+        if self.head == None:
+            self.head = Node(value, None)
+            return
+        cur = self.head
+        prev = None
+        while cur != None:
+            prev = cur
+            cur = cur.next
+        prev.next = Node(value, None)
+
 
     # method to return the max value in the linked list
     # returns the data value and not the node
     def find_max(self):
-        pass
+        if self.head == None:
+            return None
+        cur = self.head
+        cur_max = cur.value
+        cur = cur.next
+        while cur != None:
+            if cur.value > cur_max:
+                cur_max = cur.value
+            cur = cur.next
+        return cur_max
+
 
     # method to delete the first node found with specified value
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def delete(self, value):
-        pass
+        if self.head == None:
+            return None
+        if self.head.value == value:
+            self.head = self.head.next
+        
+        prev = self.head
+        cur = self.head.next
+        while cur != None and cur.value != value:
+            prev = cur
+            cur = cur.next
+        
+        if cur != None:
+            prev.next = cur.next

```

Comment:
this works! let's take away one of the variables, though. again, let's start the while loop with `cur.next != None`:

```suggestion
        cur = self.head
        while cur.next != None and cur.next.value != value:
            cur = cur.next
        cur.next = cur.next.next
```